File: Week_08/java/src/CountingSort.java
import java.util.Arrays;

public class CountingSort {
    public static void main(String[] args) {
        int[] a1 = {1, 5, 3, 2, 6, 4};
        countingSort(a1);
        System.out.println(Arrays.toString(a1));
    }

    //时间O(n + k) 空间 O(k)k表示输入的元素是0到k之间的整数
    public static void countingSort(int[] array) {
        int maxValue = findMaxValue(array);
        int[] bucket = new int[maxValue + 1];
        for (int value : array) {
            bucket[value]++;
        }

        for (int i = 0, j = 0; i < bucket.length; i++) {
            int count = bucket[i];
            while (count-- > 0) {
                array[j++] = i;
            }
        }
    }

    public static int findMaxValue(int[] array) {
        int max = array[0];
        for (int i = 1; i < array.length; i++) {
            if (array[i] > max) {
                max = array[i];
            }
        }
        return max;
    }
}

File: Week_08/cpp/src/CountingSort.cpp
#include <vector>
#include <iostream>
using namespace std;
void countingSort(vector<int> &nums)
{
    // 注意，修改nums的元素, *pmax是动态变化的
    const auto [pmin, pmax] = minmax_element(nums.begin(), nums.end());
    const int min = *pmin, max = *pmax;
    vector<int> bucket(max - min + 1, 0);
    for (int num : nums)
        bucket[num - min]++;
    int j = 0;
    for (int i = min; i <= max; i++)
    {
        int cnt = bucket[i - min];
        while (cnt-- > 0)
        {
            nums[j++] = i;
        }
    }
}
int main(int argc, char const *argv[])
{
    vector<int> nums = {1, 5, 3, 2, 2, 1, 3, 6, 4};
    countingSort(nums);
    cout << "[";
    for (int i = 0; i < nums.size(); i++)
    {
        cout << nums[i] << ",";
    }
    cout << "]" << endl;
    return 0;
}
